### ISSUE TITLE
Concurrent testing breaks snapshot assertions

### ISSUE THREAD

**Issue by NullVoxPopuli** [vitest-dev/vitest#551]
### Describe the bug

When using snapshot assertions and concurrent tests, I get this error:

```
Error: Snapshot cannot be used outside of test
```

### Reproduction

```ts
import { describe, expect, test } from 'vitest';

describe('repro', () => {
  test.concurrent('a', async () => {
    await new Promise((resolve) => setTimeout(resolve, 1000));
    expect('foo').toMatchSnapshot();
  });

  test.concurrent('b', async () => {
    await new Promise((resolve) => setTimeout(resolve, 1000));
    expect('bar').toMatchSnapshot();
  });
});

```

### System Info

```shell
System:
    OS: Linux 5.13 Ubuntu 21.10 21.10 (Impish Indri)
    CPU: (8) x64 11th Gen Intel(R) Core(TM) i5-1135G7 @ 2.40GHz
    Memory: 723.05 MB / 31.14 GB
    Container: Yes
    Shell: 5.1.8 - /bin/bash
  Binaries:
    Node: 16.13.2 - ~/.volta/tools/image/node/16.13.2/bin/node
    Yarn: 1.22.17 - ~/.volta/tools/image/yarn/1.22.17/bin/yarn
    npm: 7.24.2 - ~/Development/NullVoxPopuli/ember-apply/node_modules/.bin/npm
  Browsers:
    Chrome: 96.0.4664.93
    Firefox: 96.0.1


```
```


### Used Package Manager

yarn

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by antfu** [comment]
Yes this is the limitation of JS as you can't keep the context in async concurrent execution. I am not sure if there is a solution to that (similar to Vue's async `setup`: https://antfu.me/posts/async-with-composition-api). Otherwise we might need to document it

**Comment by sheremet-va** [comment]
> Yes this is the limitation of JS as you can't keep the context in async concurrent execution. I am not sure if there is a solution to that (similar to Vue's async `setup`: https://antfu.me/posts/async-with-composition-api). Otherwise we might need to document it

I think we can bind `test` to `expect` and pass it inside a context (first function argument), if the this idea will ever be accepted 👀

**Comment by antfu** [comment]
> I think we can bind test to expect and pass it inside a context 

Yes, there is a pending pr for it #900

**Comment by zbyte64** [comment]
Bumped into this while trying to migrate from jest.  Is there a workaround?

**Comment by ggoodman** [comment]
You might get some mileage on newer versions of node with [AsyncLocalStorage](https://nodejs.org/docs/latest-v16.x/api/async_context.html#class-asynclocalstorage)

**Comment by antfu** [comment]
It will be solved by #900

Example: https://github.com/vitest-dev/vitest/pull/900/files#diff-25788667f7dde81e3ed5b45f7e3ed4a1721820d94a9c743ceaeb6a4bada228fb

**Comment by Alioth996** [comment]
> It will be solved by #900
> 
> Example: https://github.com/vitest-dev/vitest/pull/900/files#diff-25788667f7dde81e3ed5b45f7e3ed4a1721820d94a9c743ceaeb6a4bada228fb

 It's useful
